instrumentation-core/spring/trace-core/src/main/java/io/opentelemetry/spring/instrumentation/trace/web/WebMVCFilter.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.opentelemetry.spring.instrumentation.trace.web;
+
+import io.grpc.Context;
+import io.opentelemetry.OpenTelemetry;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.context.propagation.HttpTextFormat;
+import io.opentelemetry.trace.Span;
+import io.opentelemetry.trace.Tracer;
+import io.opentelemetry.trace.TracingContextUtils;
+import java.io.IOException;
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+
+/**
+ * Wraps requests to servlet or resource in a span. Gets the parent span context from request
+ * headers.
+ */
+public final class WebMVCFilter implements Filter {
+
+  private static final HttpTextFormat.Getter<HttpServletRequest> GETTER =
```

Comment:
Can you describe strategy going forward for instrumenting more parts going forward? For example, I notice this class is duplicated from here

https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/master/instrumentation/servlet/servlet-common/src/main/java/io/opentelemetry/auto/instrumentation/servlet/HttpServletRequestGetter.java

In addition, this class itself doesn't seem to depend on spring at all.

I was sort of expecting that adding `autoconfigure` support would be a process of

- Extract core instrumentation logic out of auto-instrumentation into manual instrumentation libraries
- Add autoconfiguration / bean processors etc to recognize when an instrumented piece is being initialized and add the instrumentation

This is similar to how spring-cloud-sleuth works, which should be our reference point for adding tracing to a Spring project

https://github.com/spring-cloud/spring-cloud-sleuth/blob/master/spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/grpc/TraceGrpcAutoConfiguration.java

Having actual instrumentation live inside this module seems off, except for when instrumenting the spring library itself, so `RestTemplate` may live here.

Does this make sense?

Reply:
My plan was to instrument and autoconfigure spring webmvc and 3-4 web frameworks to start. The idea was to keep this contribution self contained and as this project matures to extract common code for manual and auto instrumentation. I noticed there is already an effort to refactor the existing auto-instrumentation. I'm not sure what the scope of that initiative is and I'm afraid I'm doing duplicate work. 

Would you recommend extracting common instrumentation code first. For example extracting some of the auto instrumentation in [apache-httpclient-4.0](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/master/instrumentation/apache-httpclient/apache-httpclient-4.0) to a new instrumentation-core:apache-httpclient:apache-httpclient-4.0 project. Then using the aspects/filters/interceptors in that newly extracted apache-httpclient-4.0 project in my instrumentation-core:spring:autoconfigure project. Is this what you originally had in mind?

I'll hold off on having instrumentation for non spring libraries in this project for now. Here's a test pr that I created with my  approach to add autoconfiguration: https://github.com/mabdinur/opentelemetry-java-instrumentation/pull/2/

Reply:
My plan moving forward is to scrap this PR and do the following: 

1. Create an instrumentation-core/springwbmvc3.1 project 
    - Move [HandlerMappingResourceNameFilter](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/master/instrumentation/spring-webmvc-3.1/src/main/java/io/opentelemetry/auto/instrumentation/springwebmvc/HandlerMappingResourceNameFilter.java) and [SpringWebMvcDecorator](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/master/instrumentation/spring-webmvc-3.1/src/main/java/io/opentelemetry/auto/instrumentation/springwebmvc/SpringWebMvcDecorator.java) to this project 
   -  I will use a HandlerMappingResourceNameFilter bean for manual/library instrumentation instead of the WebMVCFilter defined in this pull request. 
  - Add instrument('instrumentation-core:springwebmv3.1') to the instrumentation:spring-webmvc-3.1 project. 
  - Move the corresponding tests and update the new file path in different projects (Hopefully I don't introduce any regressions ;) )

2. Create instrumentation-core/spring/spring-webclients (name it `spring-core` instead?, is it better to put this project directly under instrumentation-core/?)
     - Move [RestTemplateInterceptor](instrumentation-core/spring/trace-core/src/main/java/io/opentelemetry/spring/instrumentation/trace/httpclients/RestTemplateInterceptor.java) to this project
     - Add a WebClient ExchangeFilterFunction. This will follow in the footsteps of this the TraceExchangeFilterFunction found 
       [here](https://github.com/spring-cloud/spring-cloud-sleuth/blob/master/spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/TraceWebClientBeanPostProcessor.java)
    - Update the RestTemplateInterceptor implementation to follow the decorator pattern shown in SpringWebMvcDecorator (extend class [BaseDecorator](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/master/agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/BaseDecorator.java))
 
3. Create instrumentation-core/spring/autoconfigure
    - Create RestTemplateBeanPostProcessor and WebClientBeanPostProccessor classes by implementing 
       BeanPostProccessor. Here I'll add Filters/Interceptors to RestTemplate and WebClientBuilder beans if they exist. 
   - Create a bean for spring-webmvc HandlerMappingResourceNameFilter (I'll probably rename this filter). This bean will be 
     enabled using ConditionalProperties 
  - Right now, I am focusing on spring http clients (RestTemplate and WebClient). After extracting some framworks 
    instrumentation from instrumentation/ to instrumentation-core, I will add grpc, apache http client, and okhttp beans here 
    using a similar pattern.

4. Create instrumentation-core/spring/spring-starter
  - Here I'll package the autoconfigure, spring-webclients, and the other supported frameworks. This starter will be 
    documented and hopefully published to spring.io  

5. Add more web frameworks to instrumentation-core/ and expand the auto-configuration support. 


6. This is completely unrelated but I want to provide out of the box support for basic metrics. Right now I am thinking about adding a instrumentation-core:spring:metrics-core project but I'm not sure if this is the right repository/location for this. What are your thoughts on this?

Reply:
Thanks for the details on the plan! This looks like what I had in mind too and looks like a good approach.